### ISSUE TITLE
Using Greek BERT to predict masked words

### ISSUE THREAD

**Issue by jzbjyb** [nlpaueb/greek-bert#1]
Thanks a lot for releasing the model! I was trying to use it to predict the masked words in a sentence. For example, "Είναι ένας καλός άνθρωπος." (He is a good person) and I masked out "καλός" (good). Sorry that I don't understand Greek and the above example is from Google translation, so it might be wired. However, the most plausible prediction made by Greek BERT is "##τατου", which is a little bit weird. The last dimensionality of the output tensor (which should be equal to the vocab size) is 35000, while the vocab size is 34999. Any thoughts on why this happens? BTW, I also tried multilingual BERT using the same code, and the most plausible prediction is "μόνο" (only), which seems to be reasonable.

Here is the code snippet to reproduce the example:
```python
import torch
from transformers import *

prompt = 'Είναι ένας [MASK] άνθρωπος.'
tokenizer_greek = AutoTokenizer.from_pretrained('nlpaueb/bert-base-greek-uncased-v1')
model_greek = AutoModelWithLMHead.from_pretrained('nlpaueb/bert-base-greek-uncased-v1')
inp = tokenizer_greek.encode(prompt)
print(tokenizer_greek.convert_ids_to_tokens(inp))
# ['[CLS]', 'ειναι', 'ενας', '[MASK]', 'ανθρωπος', '.', '[SEP]']
out = model_greek(torch.tensor([inp]))[0]
print(tokenizer_greek.convert_ids_to_tokens(out[0, 3].max(0)[1].item()))
# the most plausible prediction for [MASK] is "##τατου"
print(out.size(), len(tokenizer_greek.vocab))
# torch.Size([1, 7, 35000]) 34999
```

**Comment by iliaschalkidis** [comment]
Hi @jzbjyb,

**Vocabulary issue**

First of all thanks for the feedback. Given your observation, I found out that in the original vocabulary (`vocab.txt`) provided to the TF model, a special character `«` had been used in two different slots. Converting the TF model to a PyTorch one to migrate in transformers library, the second occurrence had been erased, leading to a smaller vocabulary (34999 vs. 35000), leading possibly to further misalignments between token IDs and embeddings. I re-uploaded a new vocabulary (`vocab.txt`) after replacing the second occurrence with another rare token which was not part of the initial vocabulary and I fixed the misalignment. I speculate that you have to manually delete the cached version of `nlpaueb/bert-base-greek-uncased-v1` in your workspace, in order to retrieve the newest version from AWS S3.

**Language Model issue**

Every single time I rerun your script I receive a different prediction (token). I speculate that the converted model does not include the weights of the final masked_lm layer (768,35000) and a newly created masked_lm layer is being initialised in every single run leading to a different probability distribution over the vocabulary and thus a different most probable predicted word.

This is not the case with `bert-base-multilingual-uncased`, which always predicts the same token every single time. I replaced `AutoModelWithLMHead` with `AutoModel` in order to retrieve the final output representations (N, 768) and check that these are consistently the same across various re-runs and this is the case. This means that the model uses the pre-trained weights and it is totally OK to be fine-tuned for downstream tasks. On the other hand, as I mentioned before, the `LMHead` is changing in every single run and you cannot evaluate the model as language model, which is really frustrating...

I hope my comments help you somehow...

**Comment by jzbjyb** [comment]
Hi @iliaschalkidis, thanks a lot for your quick fix and it really helps! I updated the cache and the vocab issue has been fixed. I checked the downloaded `pytorch_model.bin` file and there is indeed no parameters for the masked_lm layer. Is there any way to include this layer in the converted model?

**Comment by iliaschalkidis** [comment]
Hi @jzbjyb , thanks again for reporting. 

In relation with the **vocabulary issue**, I did some additional testing in order to validate that now the vocabulary index is aligned with the pre-trained embeddings included in the BERT model. I exported the embeddings `(35000,768)` and created a word2vec-like TXT format that I could later on load using `gensim` library. I find the the following results that show there is no further issue:

```python
from gensim.models import KeyedVectors

bert_embedding = KeyedVectors.load_word2vec_format('./greek_bert_embeddings.txt')
# Similar to 'Olympiakos' sport club,  are other famous Greek sports clubs
print([word for word, score in bert_embedding.most_similar('ολυμπιακος')])
# ['ολυμπιακο', 'ολυμπιακου', 'παναθηναικος', 'πανιωνιος', 'παοκ', 'παναθηναικου', 'ερυθρολευκ', 'αεκ', 'παναθηναικο', 'ερυθρολευκο']
# Similar to 'Tsoukala', a Greek surname, are other Greek surnames
print([word for word, score in bert_embedding.most_similar('τσουκαλα')])
# ['καφανταρη', 'κικιλια', 'κουτσουμπα', 'λαζαριδη', 'αναγνωστοπουλος', 'καραγκουνη', 'κουλουρη', 'σαββοπουλο', 'ρουβα', 'ραγκουση']
# Similar to 'good',  are variations of 'good' and 'bad'
print([word for word, score in bert_embedding.most_similar('καλος')])
# ['καλοι', 'κακος', 'καλους', 'καλες', 'καλο', 'καλη', 'καλυτερος', 'καλης', 'καλων', 'κακες']
print([word for word, score in bert_embedding.most_similar('πανεπιστημιο')])
# Similar to 'university', are variations of 'university', 'school', 'academia'
# ['πανεπιστημιου', 'πανεπιστημια', 'πανεπιστημιων', 'university', 'σχολειο', 'σχολη', 'φοιτητες', 'ακαδημαικα', 'ακαδημαικο', 'δημαρχειο']
```

Performing the same evaluation with the previous corrupted vocabulary led to completely irrelevant concepts (words). So the issue has been resolved and now everyone can download the correct vocabulary using `from_pretrained('nlpaueb/bert-base-greek-uncased-v1')`

In relation with the **language model classification layer**, I'm pretty confident that there is a work-around, because Google's BERT models loaded using transformers library do not have the same issue. As I mention in the read-me, I did the training with the official python code of Google Research. And I ended up with the following files (the first 3 are the actual weights and their mapping to the model) :

```
- variables.data-00000-of-00001
- variables.index
- variables.meta
- vocab.txt
- config.json
```

@julien-c, @thomwolf could you please assist us on how we could also port the masked LM layer weights into the `pytorch_model.bin` in order to be able to use `AutoModelWithLMHead` properly with the pre-trained weights?

**Comment by jzbjyb** [comment]
Thanks for the thorough testing! If there is anything/workaround I can do on my side please let me know!
